Support Azure MangedIdentity And WorkloadIdentity TokenCredential

priyanjgupta opened the pull request:
closes https://github.com/microsoft/garnet/issues/558

The PR also adds support for WorkloadIdentity, which will enable seamless authentication for workloads deployed on Azure Kubernetes Services (AKS)
Please refer: https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview?tabs=dotnet for more information

kevin-montrose commented on libs/host/Configuration/Options.cs:
```diff
@@ -880,4 +908,9 @@ internal enum ConfigFileType
         RedisConf = 1,
     }
+
+    public class InvalidAzureConfiguration : Exception
```
nit: should probably be `sealed`

priyanjgupta replied:
Updated, thanks for letting me know.

kevin-montrose commented on libs/storage/Tsavorite/cs/src/devices/AzureStorageDevice/BlobUtilsV12.cs:
```diff
@@ -37,6 +37,31 @@ public struct ServiceClients
+            };
+        }
+
+        private static (BlobClientOptions aggressiveOptions, BlobClientOptions defaultOptions, BlobClientOptions withRetriesOptions) GetBlobClientOptions()
```
nit: tuple member names should be PascalCased

priyanjgupta replied:
Updated, something like this?

TalZaccai commented on test/Garnet.test/GarnetServerConfigTests.cs:
```diff
@@ -208,6 +208,11 @@ public void ImportExportRedisConfigLocal()
         [Test]
         public void ImportExportConfigAzure()
         {
+            if (!TestUtils.IsRunningAzureTests)
```
Can remove `if (TestUtils.IsRunningAzureTests)` below

priyanjgupta replied:
Updated, thank you

TalZaccai commented on test/Garnet.test/GarnetServerConfigTests.cs:
```diff
@@ -243,6 +248,52 @@ public void ImportExportConfigAzure()
 
+        [Test]
+        public void AzureStorageConfiguration()
+        {
```
Please also add testing logic for these new options in `ImportExportConfigAzure` - that will verify that serializing and deserializing the configuration to/from a `garnet.conf` file is also working.

priyanjgupta replied:
Thanks for letting me know, I have added the same in ImportExportConfigAzure.
Please let me know if any further changes are needed.

TalZaccai replied:
LGTM!

TalZaccai commented:
Hey @priyanjgupta any updates on the comments from @kevin-montrose? Are those resolved?

priyanjgupta commented:
> Hey @priyanjgupta any updates on the comments from @kevin-montrose? Are those resolved?

priyanjgupta commented:
> Hey @priyanjgupta any updates on the comments from @kevin-montrose? Are those resolved?

Hey @TalZaccai I have addressed the comments. If the changes look good to you and @kevin-montrose, I can go ahead and resolve them. Let me know if I missed anything.
Thanks
